File: src/main/java/algo/twoPointers/PushDominos_838.java
package algo.twoPointers;

public class PushDominos_838 {
  public static String pushDominoes(String dominoes) {
    int len = dominoes.length();
    int start = -1, end = -1;
    var chars = new char[len + 2];

    chars[0] = 'L';
    for (int i = 0; i < len; i++) {
      chars[i + 1] = dominoes.charAt(i);
    }
    chars[len + 1] = 'R';

    // find dominoes that are not pushed
    for (int i = 0; i < chars.length; i++) {
      if (chars[i] == '.') {
        if (start == -1) {
          start = i;
        }
        end = i;
      } else {
        // now we determine the status for unpushed dominoes
        if (start != -1) {
          var left = start - 1 >=0 ? chars[start - 1] : 'L';
          var right = end + 1 < chars.length ? chars[end + 1] : 'R';
          if (left == 'L' && right == 'R') {
            // no impact to the status
          } else if (left == 'L' && right == 'L') {
            for (int j = start; j <= end; j++) {
              chars[j] = 'L';
            }
          } else if (left == 'R' && right == 'R') {
            for (int j = start; j <= end; j++) {
              chars[j] = 'R';
            }
          } else if (left == 'R' && right == 'L') {
            for (int j = start, k = end; j < k; j++, k--) {
              chars[j] = 'R';
              chars[k] = 'L';
            }
          }
        }

        start = -1;
      }
    }

    return new String(chars).substring(1, chars.length - 1);
  }

  public static void main(String[] args) {
    System.out.println(pushDominoes("RR.L"));
    System.out.println(pushDominoes(".L.R...LR..L.."));
    System.out.println(pushDominoes("L.R...RLR..L..L"));
    System.out.println(pushDominoes("L.R...RLR...L..L"));
    System.out.println(pushDominoes(".L.R."));
    System.out.println(pushDominoes("..R.."));
  }
}

File: src/main/kotlin/algo/twoPointers/PushDominos_838.kt
package algo.twoPointers

class KtPushDominos_838 {
    fun pushDominoes(dominoes: String): String {
        val states = dominoes.toMutableList()
        states.add(0, 'L')
        states.add('R')

        // find 'L' and 'R' and then change status between 'L' and 'R'
        var leftState = -1; var rightState = -1
        states.indices.forEach { i ->
            if (states[i] != '.') {
                if (leftState == -1) {
                    leftState = i
                } else {
                    rightState = i
                }
            }
            if (leftState != -1 && rightState != -1) {
                val left = states[leftState]
                val right = states[rightState]
                if (left == 'L' && right == 'L') {
                    for (k in leftState + 1..<rightState) {
                        states[k] = 'L'
                    }
                } else if (left == 'L' && right == 'R') {
                    // do nothing
                } else if (left == 'R' && right == 'L') {
                    var j = leftState; var k = rightState
                    while (j < k) {
                        states[j] = 'R'
                        states[k] = 'L'
                        j++
                        k--
                    }
                } else if (left == 'R' && right == 'R') {
                    for (k in leftState + 1..<rightState) {
                        states[k] = 'R'
                    }
                }
                leftState = rightState
                rightState = -1
            }
        }
        val s = states.joinToString("")
        return s.substring(1, s.length - 1)
    }
}

fun main() {
    val domino = KtPushDominos_838()
    println(domino.pushDominoes("RR.L"))
    println(domino.pushDominoes(".L.R...LR..L.."))
    println(domino.pushDominoes(".L.R."))
    println(domino.pushDominoes("R.R.L"))
}